### PR TITLE
GridFittedBoundary for IncompressibleModel and faster ImmersedBoundaryGrid with high-order advection

### PR DESCRIPTION
**Pull request by glwagner**
This somewhat large and WIP PR implements immersed boundary in the non-hydrostatic algorithm used by `IncompressibleModel`.

It adds two validation tests: `shallow_water_flow_past_cape.jl` and `immersed_bickley_jet.jl`. The intent of `immersed_bickley_jet.jl` is to compare simulations with a "native" boundary and an immersed boundary with the same effective domain size, grid spacing, and numerics.

@whitleyv I suggest that we merge this PR once we verify that the validation scripts run efficiently. After that we can work on careful testing and analysis of the validation case.

### CODE REVIEW

**Comment by glwagner** [comment]
After a bit of pain, things are looking ok. Here's a comparison between a "regular" bickley jet in a square domain, and a bickley jet in a slightly larger domain that's been truncated to a square by an immersed boundary on the top boundary:

https://user-images.githubusercontent.com/15271942/120709934-f8389500-c469-11eb-8e55-bce7c0659245.mp4

The color scale on the bottom panels (showing the difference between the two solutions) is 1/10th of the top.

**Comment by glwagner** [comment]
GPU tests are failing because of an issue with CUDA drivers on sverdrup but hopefully these will be resolved by tomorrow morning.

**Comment by glwagner** [comment]
256^2:

https://user-images.githubusercontent.com/15271942/120718628-0f7d7f80-c476-11eb-8962-8bcf994add2a.mp4

**Comment by glwagner** [comment]
@wenegrat might be interested to see this discussion.

**Comment by francispoulin** [comment]
If you were to impose no-flux boundary conditions on the tracer at the solid boundary then that would be one way to ensure that the total tracer is conserved.  As long as you had some diffusivity that should not be a problem, and might make things better.  However, that wouldn't solve all the problems.

**Comment by glwagner** [comment]
> If you were to impose no-flux boundary conditions on the tracer at the solid boundary then that would be one way to ensure that the total tracer is conserved. As long as you had some diffusivity that should not be a problem, and might make things better. However, that wouldn't solve all the problems.

We do impose those boundary conditions! But the velocity field doesn't satisfy impenetrable boundary conditions, eg `u . n \ne 0`. Therefore there is advection of tracer across the immersed boundary. Even if we brute force neglect that advection term, tracer still would not be conserved because the velocity field is then compressible.

**Comment by francispoulin** [comment]
Sorry that I misunderstood.  I see that you are trying to impose no-normal flow but there is some numerical error, and this error is what leads to the difficulties.  

What is the major factor that controls the size of the error that is generated?  Is it the pressure solve?  It so then maybe playing with this in `ShallowWaterModel`, where there is no pressure solve, might be insightful in terms of better enforcing the no-normal flow boundary conditions.  Maybe this would lead to some future improvements in the `IncompressibleModel`?

This is more a plan for a longer time scale than this PR though.

**Comment by wenegrat** [comment]
Tagging @weymouth for his expertise on IB methods.

My (basic) understanding of the options for getting both incompressibility and accurate boundary conditions is that you can either modify the RHS of the pressure Poisson equation, or use an iterative approach (which I think is what @glwagner suggests above). 

Either way I agree with the suggestion to merge this PR and focus on validation and improvement (also probably good to have @whitleyv add a cylinder validation case since there's lots of prior results to compare this to).

**Comment by weymouth** [comment]
I'm sort of glad to see @glwagner is verifying the issues we first discussed last year. My 2 cents: 

- Small errors in the BC eventually propagating to the whole simulation is unavoidable; this is a well mixed flow. Checking on integrated quantities, such as maintaining conserved properties or matching the expected global dissipation rate, is a more achievable and important goal. 
 - Using a test case with some stronger pressure gradients along the boundary would be good as well since this induces much of the error. The circle @wenegrat suggests is the classic, but you might be able to find a more subtle case which is more relevant to your application. Perhaps turbulent flow in a wavy-wall channel? Linking the amplitude of the waves to the change in turbulent statistics seems pretty well aligned with your application (though I could be wrong).

**Comment by wenegrat** [comment]
@weymouth Is iterating between the application of the boundary conditions and the pressure correction what you would recommend if we determine the errors are unacceptably large?

**Review by whitleyv — APPROVED** [review]
Great foundation for IBM!

**Comment by glwagner** [comment]
Thanks all for the feedback! I chose the fake wall case because it permits a direct, comparison between a "true" solution and the immersed boundary solution. But I agree that judging the fidelity of statistics and budgets in a more complicated case is also necessary. The challenge of more complicated cases is finding reference statistics to compare to. It'd be nice to find an "[Askervein Hill](https://news.ucar.edu/126732/capturing-detailed-portrait-wind) of the ocean"...

As for modifications to the pressure solver, we may be able to use an preconditioned conjugate gradient method (an iterative, matrix-free method) with the FFT-based solver as a preconditioner. I think a conjugate gradient method could provide some improvement over a "naive" iterative approach (iteratively applying the pressure correction step). We've already coded up such a solver to work with Oceananigans data structures:

https://github.com/CliMA/Oceananigans.jl/blob/master/src/Solvers/preconditioned_conjugate_gradient_solver.jl

A multigrid method would also be nice and might be even faster... ! But, probably a bit more difficult to code up I think. Getting the conjugate gradient solver to work might not take long.

**Comment by francispoulin** [comment]
I have never used multigrid solvers in julia but did a quick look around and found there are at least a couple of libraries one can use.  If someone decides they want to play with multigrid, I would suggest looking at these and other libraries as that would be very fast to get started.

https://github.com/JuliaLinearAlgebra/AlgebraicMultigrid.jl
https://github.com/JuliaInv/Multigrid.jl

**Comment by weymouth** [comment]
I looked at both of those and they aren't very fast. That's why I coded up
my own in WaterLily.

I have used CG preconditioners before with variable coefficient Poisson
equations and I agree they are a nice matrix free approach.

Gabriel D Weymouth
_______________________________________________
"Computers are useless. They can only give you answers." Pablo Picasso

**Comment by navidcy** [comment]
@glwagner is this ready? :)

**Comment by glwagner** [comment]
It's up to @whitleyv ! I think it'd be appropriate to compute a few more metrics (such as the tracer budget), but this could be done here or in a new PR.

**Comment by whitleyv** [comment]
> It's up to @whitleyv ! I think it'd be appropriate to compute a few more metrics (such as the tracer budget), but this could be done here or in a new PR.

I have a few metrics I've done recently. I'll add some results here, and we can discuss before merging!

**Comment by whitleyv** [comment]
I ran a few tests using the [Bickley Jet](https://github.com/CliMA/Oceananigans.jl/blob/glw-vw/grid-fitted-incompressible-model/validation/immersed_boundaries/immersed_bickley_jet.jl) with an immersed top wall from the validation script. 

### Surface Normal Velocity
We can easily look at the surface normal velocity of the immersed wall with increased resolution and see that it is converging nicely, as this method should be exact except for the pressure correction, which will scale.

Bickley_normal

### Tracer Conservation
We can also look at tracer conservation. The initial concentration is sinusoidal, so the area integrated concentration is zero over the domain initially. I've plotted both the IBM and non-IBM (lighter colors) results to see the differences. This one is on a log plot. You can see that leakage is happening in the IBM versions. It's hard to say how bad the leakage is here, and it'll probably be better to consider an initial constant concentration rather than sinusoidal to see how much leakage is occurring due to the immersed solid.

volint_Concentration_log

It might be easier to see what is going on in the IBM cases without the log scaling:

volint_Concentration

### Integrated Boundary Stress
Finally, I looked at the integrated boundary stress along the top wall. The error between the nonIBM and IBM versions at these same grid sizes are below. Unfortunately this is not getting better with increased resolution. This could be due to base state changes causing the stress to change and not necessarily something wrong with the method. It might take better refinement than 256 x 256 to really see a trend here.

Bickley_dudy_diff

Any thought on these results would be greatly appreciated, or if there is something else that might be good to look at with this particular test case. Further refinements to the grid are taking quite a long time to run on my PC, but they can be done.

These were run with a viscosity of 0 for the isotropic closure. I want to look at the viscous cases of these to see if the results change at all with added viscosity. I'm also working on some cylinder validation metrics for this IBM, but we can move forward without those for now if we like what we're seeing here!

**Comment by glwagner** [comment]
Should we add a second tracer with constant concentration to the simulation (basically measures the "mass" that's lost)?

**Comment by francispoulin** [comment]
These resutls are great @whitleyv , well done!  A few thoughts.

First, if `c` is the concentration of a tracer it should be non-negative.  If you are picking it to be a `sine` I might suggest having `1 + sin` just to avoid negative values. That being said, I don't think it's going to have any impact on the results but might be worth trying.

Second, are there any boundary conditions imposed on the tracer at the immersed boundary? I remember there was a discussion but I don't remember the conclusion.  Sorry.

Third, what exactly is computed in the last plot fo the integrated stress tensor?  Also, have you computed these for the nonIBM case to see what the real wall does?

**Comment by glwagner** [comment]
The xz and yz components of stress (called `viscous_flux_uz` and `viscous_flux_vz` in the code) are explicitly elided in the presence of an immersed boundary. For examle:

https://github.com/CliMA/Oceananigans.jl/blob/55500496bc76e5870a0e455d1dda059793e827d3/src/ImmersedBoundaries/grid_fitted_immersed_boundary.jl#L59

where

https://github.com/CliMA/Oceananigans.jl/blob/55500496bc76e5870a0e455d1dda059793e827d3/src/ImmersedBoundaries/grid_fitted_immersed_boundary.jl#L45

Since we can prescribe stresses / fluxes explicitly for grid fitted schemes (in principle --- its not implemented yet) I don't think we need to worry about errors in prescribed boundary fluxes. This issue may need to be revisited for other immersed boundary methods if they make approximations there.

File: validation/immersed_boundaries/immersed_bickley_jet.jl [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+                                           x = (-2π, 2π), y=(-2π, 3π),
+                                           topology = (Periodic, Bounded, Flat))
+
+    immersed_grid = ImmersedBoundaryGrid(expanded_grid, GridFittedBoundary(solid))
+
+    immersed_model = IncompressibleModel(architecture = arch,
+                                         advection = advection,
+                                         timestepper = :RungeKutta3,
+                                         grid = immersed_grid,
+                                         tracers = :c,
```
Comment (glwagner) [inline validation/immersed_boundaries/immersed_bickley_jet.jl:64]:
```suggestion
                                         tracers = (:c, :mass),
```

File: validation/immersed_boundaries/immersed_bickley_jet.jl [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+    ℓ = 0.5 # Gaussian width
+    k = 0.5 # Sinusoidal wavenumber
+
+    # Total initial conditions
+    uᵢ(x, y, z) = U(y) + ϵ * ũ(x, y, ℓ, k)
+    vᵢ(x, y, z) = ϵ * ṽ(x, y, ℓ, k)
+    cᵢ(x, y, z) = C(y, grid.Ly)
+
+    set!(regular_model, u=uᵢ, v=vᵢ, c=cᵢ)
+    set!(immersed_model, u=uᵢ, v=vᵢ, c=cᵢ)
```
Comment (glwagner) [inline validation/immersed_boundaries/immersed_bickley_jet.jl:85]:
```suggestion
    set!(immersed_model, u=uᵢ, v=vᵢ, c=cᵢ, mass=1)
```

**Comment by glwagner** [comment]
> Second, are there any boundary conditions imposed on the tracer at the immersed boundary? I remember there was a discussion but I don't remember the conclusion. Sorry.

Boundary conditions are zero _second-order_ / diffusive flux on tracer. However, there can be non-zero advective flux as noted.

**Comment by whitleyv** [comment]
> Third, what exactly is computed in the last plot fo the integrated stress tensor? Also, have you computed these for the nonIBM case to see what the real wall does?

Sorry probably could have been clearer on this. This is just du/dy, calculated with a centered difference between the closest two fluid cells to the boundary, then integrated along the boundary line (ie. du/dy summed and multiplied by Lx). The plot above is the error between the IBM and nonIBM case. This plot here is the values themselves (dashed is nonIBM, solid is IBM)
Bickley_dudy
 
> > First, if `c` is the concentration of a tracer it should be non-negative. If you are picking it to be a `sine` I might suggest having `1 + sin` just to avoid negative values. That being said, I don't think it's going to have any impact on the results but might be worth trying.
> 
> Fair point; using `1 + sin(x)` as the initial condition would measure leakage since the total "mass" is then Lx * Ly. I propose adding another tracer though specifically for that purpose since it's easy.

You are correct. I think the added tracer idea would be perfect. My mind was stuck on the cylinder concentration one I've done before, and didn't think about the difference here!

File: validation/immersed_boundaries/immersed_bickley_jet.jl (hunk 2 above)
Comment (whitleyv) [reply validation/immersed_boundaries/immersed_bickley_jet.jl:85]: I know in this model it doesn't matter as much, but to simulate the wall accurately it might be better to set the initial condition to be zero within the solid, just to be consistent for other IBM's of the future.

**Comment by whitleyv** [comment]
> > First, if `c` is the concentration of a tracer it should be non-negative. If you are picking it to be a `sine` I might suggest having `1 + sin` just to avoid negative values. That being said, I don't think it's going to have any impact on the results but might be worth trying.
> 
> Fair point; using `1 + sin(x)` as the initial condition would measure leakage since the total "mass" is then Lx * Ly. I propose adding another tracer though specifically for that purpose since it's easy.

I ran a tracer with initial value of 1 everywhere to look at the concentration leakage. I took the percentage of the difference in the area integrated concentration of the IBM and nonIBM scaled by the initial concentration, ie.  `100 x  abs(IBM - nonIBM)/ initial`. The results are below. It looks like for a 256 x 256 grid after 200s only about 3 x 10^-5 % of the initial concentration has leaked out, which is pretty good and seems to scale well with refinement!
volint_Concentration_leakage

File: validation/immersed_boundaries/immersed_bickley_jet.jl (hunk 2 above)
Comment (glwagner) [reply validation/immersed_boundaries/immersed_bickley_jet.jl:85]: We are applying the mask in `set!` right now:

https://github.com/CliMA/Oceananigans.jl/blob/1775ebb75bfc6beeef8b3c147223588af4488326/src/Models/IncompressibleModels/set_incompressible_model.jl#L44-L47

so there's no need to require this of users. Do you think this is a good approach / convenient for users?

Feel free to commit my suggested changes (or reject them!) plus all of the awesome code you've written for the analysis (which doesn't appear to be part of the PR yet?)

**Comment by glwagner** [comment]
> which is pretty good and seems to scale well with refinement!

Why do we expect less leakage as the grid is refined?

**Comment by glwagner** [comment]
> The plot above is the error between the IBM and nonIBM case.

Ok, I understand now. So the error in the wall-normal velocity gradient is a proxy for the error that _would_ result were we to prescribe a no-slip boundary condition with non-zero viscosity.

**Comment by whitleyv** [comment]
> > which is pretty good and seems to scale well with refinement!
> 
> Why do we expect less leakage as the grid is refined?

I believe for your implementation, since the  tracer values are not corrected for pressure, any leakage will be due to the advection of tracer by momentum that _is_ impacted by the pressure correction. This advection should be less with a smaller spacial step if I remember correctly.

File: validation/immersed_boundaries/immersed_bickley_jet.jl (hunk 2 above)
Comment (whitleyv) [reply validation/immersed_boundaries/immersed_bickley_jet.jl:85]: Gotcha okay that makes sense. It is probably easier for users not to have to deal with that! And I haven't added it yet, because it was just in a notebook and not a validation script until today!

**Comment by glwagner** [comment]
> I believe for your implementation, since the tracer values are not corrected for pressure, any leakage will be due to the advection of tracer by momentum that _is_ impacted by the pressure correction. This advection should be less with a smaller spacial step if I remember correctly.

Well, I agree that less leakage implies that less tracer is advected across the boundary. But why is less tracer advected across the boundary?

It occurs to me that we are not converging to a particular solution as we refine the grid in this case, because this problem has no viscosity. Perhaps we should do a convergence test for a case with finite viscosity.

**Comment by francispoulin** [comment]
Lots of interesting questions!

I think that doing a test with diffusivity of tracer, and maybe viscosity, would be interesting.  In that limit we would impose no-flux boundary conditions at the walls (and immersed walls) and I would think that conservation should be better in both cases.

**Comment by weymouth** [comment]
> It occurs to me that we are not converging to a particular solution as we refine the grid in this case, because this problem has no viscosity. Perhaps we should do a convergence test for a case with finite viscosity.

This is likely to help since the velocity is small in the boundary layer, and so will reduce the magnitude of the local correction in a step.

> We do introduce grid-scale gradients in the predictor velocity field via masking, so it does seem possible to me that the leakage / pressure gradient error might scale with resolution. It'd be nice to have a solid mathematical explanation for this behavior.

I would be interested in such an argument. Other than zooming in on the slowest part of the BL, I'm not sure if/why increasing resolution would reduce the total flux over a fixed interval of time.

File: validation/immersed_boundaries/immersed_bickley_jet.jl [hunk 3]
```diff
@@ -1,9 +1,12 @@
+pushfirst!(LOAD_PATH, joinpath(@__DIR__, "..", ".."))
```
Comment (glwagner) [inline validation/immersed_boundaries/immersed_bickley_jet.jl:1]: <3
